lib/components/user/user-account-screen.js

```diff
@@ -0,0 +1,243 @@
+        const userData = isNewAccount ? getStateForNewUser(auth.user) : result.data
+        if (userData.savedLocations === null) userData.savedLocations = []
+
+        this.setState({
+          isNewAccount,
+          userData
+        })
+      }
+    } catch (error) {
+      // TODO: improve error handling.
+      alert(`An error was encountered:\n${error}`)
+    }
+  }
+
+  _updateUserState = newUserData => {
+    const { userData } = this.state
+    this.setState({
+      userData: {
+        ...userData,
+        ...newUserData
+      }
+    })
+  }
+
+  _updateUserPrefs = async () => {
+    const { auth, persistence } = this.props
+
+    if (persistence && persistence.otp_middleware) {
+      const { accessToken } = auth
+      const { isNewAccount, userData } = this.state
+
+      // TODO: Change state of Save button.
+
+      let result
+      if (isNewAccount) {
+        result = await addUser(persistence.otp_middleware, accessToken, userData)
+      } else {
+        result = await updateUser(persistence.otp_middleware, accessToken, userData)
+      }
+
+      // TODO: improve this.
+      if (result.status === 'success') {
+        alert('Your preferences have been saved.')
+      } else {
+        alert(`An error was encountered:\n${result.message}`)
+      }
+    }
+  }
+
+  _handleExit = () => {
+    const { originalUrl } = this.props
+    this.props.routeTo(originalUrl)
+  }
+
+  _handleExitAndSave = async () => {
+    await this._updateUserPrefs()
+    this._handleExit()
+  }
+
+  /**
+   * Hook userData, onUserDataChange on some panes upon rendering.
+   * This returns a new render function for the passed component
+   * that allows passing other props to it later if needed.
+   */
+  _hookUserData = Pane => props => {
+    const { userData } = this.state
+    return (
+      <Pane
+        onUserDataChange={this._updateUserState}
+        userData={userData}
+        {...props}
+      />
+    )
+  }
+
+  // Make an index of pane components, so we don't render all panes at once on every render.
+  // Hook some panes the userData and onUserDataChange props.
+  _panes = {
+    terms: this._hookUserData(TermsOfUsePane),
+    notifications: this._hookUserData(NotificationPrefsPane),
+    verifyPhone: PhoneVerificationPane,
+    locations: this._hookUserData(FavoriteLocationsPane),
+    finish: AccountSetupFinishPane
+  }
+
+  async componentDidMount () {
+    await this._fetchUserData()
+
+    // TODO: Update title bar.
+  }
+
+  render () {
+    const { auth } = this.props
+    const { user } = auth
+    const {
+      isNewAccount,
+      userData
+    } = this.state
+
+    if (!userData) {
+      // Display a hint while fetching user data (from componentDidMount).
+      return <AwaitingScreen />
+    }
+
+    return (
+      <div className='otp'>
+        {/* TODO: Do mobile view. */}
+        <AppNav />
+        <form className='container'>
+          {(isNewAccount
```

Comment:
This is a lot of nesting. I'd prefer for it to be simplified by having something like:

```js
let formContents
if (!isNewAccount) {
  formContents = (
    <ExistingAccountDisplay
      onCancel={this._handleExit}
      onComplete={this._handleExitAndSave}
      panes={this._panes}
    />
  )
} ...

return (
...
  <form className='container'>
    {formContents}
  </form>
...
)
```